### PR TITLE
Pr/ci add icx

### PR DESCRIPTION
**Pull request by AlexMWells**
## Description

Add compiler support for Intel(r) oneAPI DPC++/C++ Compiler (based on LLVM technology) aka icx.

Detect if CMAKE_CXX_COMPILER_ID matches "IntelLLVM" and set CMAKE_COMPILER_IS_INTELCLANG=1
NOTE:  CMAKE_COMPILER_IS_INTEL will not be set, that is reserved for the classic Intel(r) C++ Compiler.

Define OSL_INTEL_LLVM_COMPILER to hold an encoded Intel(r) LLVM Compiler version (e.g. 20220000), or 0 if not an Intel(r) LLVM Compiler.
Define OSL_INTEL_CLANG_VERSION to hold the encoded Clang version the Intel(r) LLVM Compiler is based on (e.g. 140000),
or 0 if not an Intel(r) LLVM compiler.

Ensure OSL_CLANG_VERSION is when its a vanilla clang compiler, not apple or Intel(r) oneAPI DPC++/C++ Compiler.

Added OSL_ANY_CLANG which is set for any clang based compiler.

Removed OSL_NON_CLANG_COMPILER.
Updated uses of OSL_NON_CLANG_COMPILER with more specific and understandable combinations of OSL_ANY_CLANG, OSL_INTEL_CLANG_VERSION, and OSL_INTEL_COMPILER.

Make sure our math calculations are consistent, especially division.  Different compilers may have choose to use reciprocal division loosing precision causing slightly different results which can lead to aliasing differences when running the testsuite.
Disabled reciprocal division for all projects.

if -ffast-math is enabled to to the underlying compiler, inf's and NaN's may not be handled properly (by design for preformance).  When building liboslexec, ensure any source files that require proper INF or NaN handling have additional command line flags passed if needed (as the default behavior maybe -ffast-math):
    shadingsys.cpp
    wide/wide_shadingsys
    wide/wide_optest_float

In in impl_transform_normal_masked, enable unitialized Matrix44 because all of its members get overwritten anyway, 

## Tests

Added icx-vp2022 to CI plan to exercise icx.

Added alternative test results where needed, using '.icx.' as shorthand for the Intel(r) oneAPI DPC++/C++ Compiler in alternative test result filenames.
Changed some tests to use the pretty(value) which clamps values close to 0 to be 0 to avoid lots of different alternative rest result files when using printf.
Added "--center" option to several regression tests to avoid aliasing in results between SIMD batched and scalar execution.

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] I have previously submitted a [Contributor License Agreement](http://opensource.imageworks.com/cla/).
- [x] I have updated the documentation, if applicable.
- [x] I have ensured that the change is tested somewhere in the testsuite (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

### CODE REVIEW

File: CMakeLists.txt
Comment (lgritz) [inline CMakeLists.txt:180]:
```suggestion
# reciprocal division losing precision causing slightly different
```

File: CMakeLists.txt
Comment (lgritz) [inline CMakeLists.txt:190]: Does this have performance implications we should be aware of? Is this something we should do all the time, or just for CI tests for the sake of uniformity of results across compilers?

**Review by lgritz — APPROVED** [review]
Great!

File: CMakeLists.txt
Comment (AlexMWells) [reply CMakeLists.txt:190]: Yes it has performance implications (as a reciprocal divide instruction combined with a multiply is faster).
Yes its just for sake of CI tests and the sake of uniformity of results across compilers!
There is a difference in default behavior among compilers, and hardware availability for the reciprocal divide.

For performance, I would do the reverse, explicitly enable reciprocal divides.  However precision will be lost, perhaps more than a renderer would want. 

As a programmer, I feel that the programmer probably should of written (1/divisor)*number instead of number/divisor if their algorithm would be OK with the loss in precision.

Comment (AlexMWells) [reply CMakeLists.txt:190]: Also tests themselves could be updated to be more resilient (example is using --center) and alternate results could be added as well.  FMA's have similar issue but actually increase precision/correctness of results vs. a reciprocal divide which looses precision.

File: .github/workflows/ci.yml
Comment (lgritz) [inline .github/workflows/ci.yml:556]: I believe the `FMT_VERSION` and `OPENIMAGEIO_CMAKE_FLAGS` lines are not necessary for icx, only icc.

File: .github/workflows/ci.yml
Comment (lgritz) [inline .github/workflows/ci.yml:561]: I had commented out the `USE_SIMD` for icc because it uses a different syntax than the `-m` arguments that this gets turned into by compiler.cmake. But icx accepts the same `-m` arguments as clang. So you can re-enable that line for icx.

Comment (lgritz) [inline .github/workflows/ci.yml:561]: I had commented out the `USE_SIMD` for icc because it uses a different syntax than the `-m` arguments that this gets turned into by compiler.cmake. But icx accepts the same `-m` arguments as clang. So you can re-enable that line for icx.

File: src/build-scripts/gh-installdeps.bash
Comment (lgritz) [inline src/build-scripts/gh-installdeps.bash:52]: I think these lines (printing the versions) are wholly unnecessary, since early in the "build" stage, the cmake scripts print the compiler and version, so it's already easy to verify looking at the logs. (In fact, these lines can be misleading, since it only tells you that an icpc or icpx *exist*, but not what we actually are building with it!)

Comment (lgritz) [inline src/build-scripts/gh-installdeps.bash:52]: I think these lines (printing the versions) are wholly unnecessary, since early in the "build" stage, the cmake scripts print the compiler and version, so it's already easy to verify looking at the logs. (In fact, these lines can be misleading, since it only tells you that an icpc or icpx *exist*, but not what we actually are building with it!)

File: src/build-scripts/gh-installdeps.bash
Comment (lgritz) [inline src/build-scripts/gh-installdeps.bash:100]: Unnecessary -- see other note

Comment (lgritz) [inline src/build-scripts/gh-installdeps.bash:100]: Unnecessary -- see other note

File: src/include/OSL/platform.h
Comment (lgritz) [inline src/include/OSL/platform.h:89]: Are both of these necessary? For the other compilers, we only define OSL_COMPILERNAME_VERSION, set to some encoding of the version, and to 0 if that is not the compiler being used. If `#if OSL_INTEL_LLVM_COMPILER` isn't every going to evaluate differently than `#if OSL_INTEL_CLANG_VERSION`, then let's just eliminate the former one. (I don't have a preference for whether you call it OSL_INTEL_CLANG_VERSION or OSL_INTEL_LLVM_VERSION... pick whichever you think is better, and I'll change the OIIO side to conform.)

Comment (lgritz) [inline src/include/OSL/platform.h:89]: Are both of these necessary? For the other compilers, we only define OSL_COMPILERNAME_VERSION, set to some encoding of the version, and to 0 if that is not the compiler being used. If `#if OSL_INTEL_LLVM_COMPILER` isn't every going to evaluate differently than `#if OSL_INTEL_CLANG_VERSION`, then let's just eliminate the former one. (I don't have a preference for whether you call it OSL_INTEL_CLANG_VERSION or OSL_INTEL_LLVM_VERSION... pick whichever you think is better, and I'll change the OIIO side to conform.)

File: src/include/OSL/wide.h
Comment (lgritz) [inline src/include/OSL/wide.h:3142]: Aside: should we consider renaming `OSL_INTEL_COMPILER` to `OSL_INTEL_CLASSIC_COMPILER` just to clarify that it only means icc, and not "any of the Intel compilers"? (I found myself trying to re-remind myself that constantly as I read this code and worked on the OIIO side.)

Comment (lgritz) [inline src/include/OSL/wide.h:3142]: Aside: should we consider renaming `OSL_INTEL_COMPILER` to `OSL_INTEL_CLASSIC_COMPILER` just to clarify that it only means icc, and not "any of the Intel compilers"? (I found myself trying to re-remind myself that constantly as I read this code and worked on the OIIO side.)

File: src/liboslnoise/wide/wide_gabornoise.h
Comment (lgritz) [inline src/liboslnoise/wide/wide_gabornoise.h:44]: Aside, that we don't need to address in this PR:

There are a lot of instances of this construct that means "if we trust omp simd for complex loops". If a future clang is better with these, we will need to not only change this in quite a few locations, but also it'll need to be a more complicated test that checks which specific version of which clang compilers.

Perhaps it is better to JUST ONCE, in platform.h, have a single

```
#if OSL_OPENMP_SIMD && (OSL_GCC_VERSION || OSL_INTEL_COMPILER || OSL_INTEL_CLANG_VERSION || ...etc...)
    #define OSL_OMP_SIMD_COMPLEX_LOOP OSL_PRAGMA(omp simd simdlen(__OSL_WIDTH))
#else
    #define OSL_OMP_SIMD_COMPLEX_LOOP
#endif
```

and then all these spots can simply say

```
    OSL_OMP_SIMD_COMPLEX_LOOP
    for (...)
```

So in a future where certain particular versions of clang are known to handle these loops well, then we can modify the single master definition and have all the loops do the right thing with no change of cut-and-paste errors or missing any.

Comment (lgritz) [inline src/liboslnoise/wide/wide_gabornoise.h:44]: Aside, that we don't need to address in this PR:

There are a lot of instances of this construct that means "if we trust omp simd for complex loops". If a future clang is better with these, we will need to not only change this in quite a few locations, but also it'll need to be a more complicated test that checks which specific version of which clang compilers.

Perhaps it is better to JUST ONCE, in platform.h, have a single

```
#if OSL_OPENMP_SIMD && (OSL_GCC_VERSION || OSL_INTEL_COMPILER || OSL_INTEL_CLANG_VERSION || ...etc...)
    #define OSL_OMP_SIMD_COMPLEX_LOOP OSL_PRAGMA(omp simd simdlen(__OSL_WIDTH))
#else
    #define OSL_OMP_SIMD_COMPLEX_LOOP
#endif
```

and then all these spots can simply say

```
    OSL_OMP_SIMD_COMPLEX_LOOP
    for (...)
```

So in a future where certain particular versions of clang are known to handle these loops well, then we can modify the single master definition and have all the loops do the right thing with no change of cut-and-paste errors or missing any.

**Comment by lgritz** [comment]
If you would like to test your osl-side build CI against my in-review OIIO-side changes, you can add these lines to the workflows/ci.yml in the `env` section of the icx test (and remove it before we're ready for the final merge):

      OPENIMAGEIO_REPO: lgritz/oiio
      OPENIMAGEIO_VERSION: lg-icx

That will cause the "build_openimageio.bash" script used by the CI setup to pull MY branch instead of the master of the official repo.

**Comment by lgritz** [comment]
If you would like to test your osl-side build CI against my in-review OIIO-side changes, you can add these lines to the workflows/ci.yml in the `env` section of the icx test (and remove it before we're ready for the final merge):

      OPENIMAGEIO_REPO: lgritz/oiio
      OPENIMAGEIO_VERSION: lg-icx

That will cause the "build_openimageio.bash" script used by the CI setup to pull MY branch instead of the master of the official repo.

File: CMakeLists.txt
Comment (AlexMWells) [reply CMakeLists.txt:180]: ok

File: .github/workflows/ci.yml
Comment (AlexMWells) [reply .github/workflows/ci.yml:556]: good catch, I'll try removing those.

File: .github/workflows/ci.yml
Comment (AlexMWells) [reply .github/workflows/ci.yml:561]: I'll try adding USE_SIMD: avx2,f16c back in

File: src/build-scripts/gh-installdeps.bash
Comment (AlexMWells) [reply src/build-scripts/gh-installdeps.bash:52]: Up to you, we can remove them entirely or add more context
```suggestion
        if [[ "$CXX" == "icpc" || "$CC" == "icc" || "$USE_ICC" != "" ]] ; then
            echo Verify installation of Intel(r) C++ Compiler
            icpc --version
        fi
        if [[ "$CXX" == "icpx" || "$CC" == "icx" || "$USE_ICX" != "" ]] ; then
            echo Verify installation of Intel(r) oneAPI DPC++/C++ Compiler
            icpx --version
        fi
```

File: src/include/OSL/platform.h
Comment (AlexMWells) [reply src/include/OSL/platform.h:89]: Depends on what you want to check on.  If you have a workaround that is more clang version based, then it might be nice to check the clang version, instead of having to figure out which __INTEL_LLVM_COMPILER version's map to which clang versions (20220000 __INTEL_LLVM_COMPILER  maps to clang version 140000.  If only 1, then OSL_INTEL_LLVM_COMPILER as that might rev several times without changing the underlying clang compiler.

File: src/include/OSL/wide.h
Comment (AlexMWells) [reply src/include/OSL/wide.h:3142]: For OSL, I would be fine (prefer) OSL_INTEL_CLASSIC_COMPILER, but do we keep OSL_INTEL_COMPILER around (comment it deprecated)?

File: src/liboslnoise/wide/wide_gabornoise.h
Comment (AlexMWells) [reply src/liboslnoise/wide/wide_gabornoise.h:44]: The OSL_NON_INTEL_CLANG always bugged me when reading it, and it got worse with the new mix of classic and llvm based compilers.  So I would just go ahead and make this change now and feel better about it!

Comment (AlexMWells) [reply src/liboslnoise/wide/wide_gabornoise.h:44]: I still like to pass arguments vs. assuming simdlen(__OSL_WIDTH), but perhaps something like 
```
#define OSL_OMP_SIMD_LOOP(...) OSL_OMP_PRAGMA(omp simd __VA_ARGS__)
#if (OSL_GCC_VERSION || OSL_INTEL_COMPILER)
    #define OSL_OMP_COMPLEX_SIMD_LOOP(...) OSL_OMP_PRAGMA(omp simd __VA_ARGS__)
#else
     // Ignore requests to vectorize complex SIMD loops for certain compiler's/versions
    #define OSL_OMP_COMPLEX_SIMD_LOOP(...)
#endif
```

File: .github/workflows/ci.yml
Comment (lgritz) [reply .github/workflows/ci.yml:561]: It worked for me on the OIIO side.

I've been contemplating a change to the logic in compiler.cmake to basically be sure to do the right thing to the USE_SIMD for icc under the covers (i.e., if it sees USE_SIMD=avx512f, instead of just passing on `-mavx512f`, which I think is not recognized by icc, it would translates to `-axCORE-AVX512` [I think]). But now that I see that icx recognizes all the same commands as clang, I'm wondering if extra work for icc is really worth the trouble (or if users should just use EXTRA_CPP_ARGS to set the right thing if they know it) if icc users are going to tend to switch to icx over time anyway.

refs:
https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/code-generation-options/m.html
https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/code-generation-options/x-qx.html

File: src/include/OSL/platform.h
Comment (lgritz) [reply src/include/OSL/platform.h:89]: Oh, I see what you mean. Are you saying that `OSL_INTEL_LLVM_COMPILER` holds the actual Intel version number, while `OSL_INTEL_CLANG_VERSION` holds the version of the clang that it's emulating or derived from?

I misunderstood, and thought you were using OSL_INTEL_LLVM_COMPILER simply as a bool of sorts.

So, in that case, let me change my recommendation. Let's change  `OSL_INTEL_LLVM_COMPILER` to `OSL_INTEL_LLVM_COMPILER_VERSION` to make clear that it's a version number, not just a bool, and thus match `OSL_GNUC_VERSION`, `OSL_APPLE_CLANG_VERSION`, etc.

File: src/include/OSL/wide.h
Comment (lgritz) [reply src/include/OSL/wide.h:3142]: On one hand, this is a public header, so that would argue for keeping the old one as a (deprecated) alias until we're moving to a 2.0 release where we're allowed to break user source code. On the other hand, this preprocessor symbol is for us and it's not a documented part of the public APIs?

Your call. I don't feel strongly either way in this particular case.

File: .github/workflows/ci.yml
Comment (AlexMWells) [reply .github/workflows/ci.yml:561]: Not a make file guru, if passing native flags, then just make user pass them in the 1st place.  Or have those be OSL defined and controlled flags that get translated.  Technically adding "-m" does count as translating, so having it spit out "-axCORE-AVX512" seems totally reasonable.  You do know the "-ax" is automatic dispatch that generates multiple versions of functions with a spring board to jump to the correct one based on CPUID?

**Comment by lgritz** [comment]
> If you would like to test your osl-side build CI against my in-review OIIO-side changes, you can add these lines to the workflows/ci.yml in the `env` section of the icx test (and remove it before we're ready for the final merge):

Scratch that. I've merged the changes now, so current OIIO master should give you a clean compile with icx.

File: src/include/OSL/platform.h
Comment (AlexMWells) [reply src/include/OSL/platform.h:89]: Seems valid reasoning.  This means the older 
OSL_INTEL_COMPILER
which also has the compiler version (its not just a bool), should be renamed as well.  And you were already suggesting
OSL_INTEL_CLASSIC_COMPILER, so should it be 
OSL_INTEL_CLASSIC_COMPILER_VERSION
?

File: src/include/OSL/wide.h
Comment (AlexMWells) [reply src/include/OSL/wide.h:3142]: Although a compilation error would be annoying, If downstream project is using OSL_INTEL_COMPILER, they should get a nice compile error and easily update it on their end.  So, I'll just drop it.

File: src/liboslnoise/wide/wide_gabornoise.h
Comment (lgritz) [reply src/liboslnoise/wide/wide_gabornoise.h:44]: I think that's reasonable, yes.

**Comment by lgritz** [comment]
I see this is passing all the tests now! Is it ready for review/merge?

**Comment by AlexMWells** [comment]
Yes, its ready for review/merge.

**Pull request merged**
